src/core/node.cc

```diff
@@ -104,6 +121,16 @@ void AbstractNode::progress_report() const
   progress_update(shared_from_this(), this->progress_mark);
 }
 
+std::shared_ptr<AbstractNode> AbstractNode::clone() const {
+  std::shared_ptr<AbstractNode> n = this->cloneOne();
+  n->children.clear();
+  n->idx = idx_counter++;
+  for (auto child : children) {
```

Comment:
seems copying of shared_ptr is redundant here

Reply:
It's been almost six months since I last looked at this and so maybe my brain is just fuzzy, but which copying do you mean?  The fact that we clone the AbstractNode rather than just taking another pointer to it?  If so, note `n->idx`; I believe that I tried just taking another pointer but found that it was critical that each node in the tree have a unique `n->idx`.

Reply:
`auto child` creates copy of element from `children` which is vector of shared pointers

Reply:
[ I should note that my C++ is really quite weak, and so I may easily miss things that should be obvious. ]

So I think the copying that you refer to is the copying of the pointer itself, with the associated bump in the reference count, and that you would suggest instead `for (const auto& child: children)`.  Right?

Reply:
Yes, you are right. Copying shouldn't be a significant performance problem, I suppose, but if we can improve it, why we shouldn't